activesupport/lib/active_support/cache/streaming_compressor.rb

```diff
@@ -0,0 +1,110 @@
+# frozen_string_literal: true
+
+require "zlib"
+require "concurrent/executor/cached_thread_pool"
+require "active_support/core_ext/numeric/bytes"
+
+module ActiveSupport
+  module Cache
+    # Serializes object and compresses serialized representation in streaming way.
+    # This avoids storing whole copy of serialized intermediate representation,
+    # like this happens in `Zlib.deflate(Marshal.dump(value))`.
+    module StreamingCompressor
+      DEFLATE_BUFFER_SIZE = 16.kilobytes.to_i
+      INFLATE_CHUNK_SIZE = 4.kilobyte.to_i
+
+      extend self
+
+      # Dumps value into string. Returns nil if `:compress_threshold` option
+      # is set and serialized value is less than given value.
+      def dump(value, **options)
+        io = StringIO.new
+        raw_bytes = dump_to_io(value, io, **options)
+        return unless raw_bytes
+        io.rewind
+        result = io.read
+        result if result.bytesize < raw_bytes
+      end
+
+      # Same as dump but streams compressed data. Returns nil and don't stream
+      # anything if `:compress_threshold` option is set
+      # and serialized value is less than given value.
+      def dump_to_io(value, to_io, compress_threshold: nil)
+        rio, wio = binary_pipe
+        within_safe_thread do
+          begin
+            Marshal.dump(value, wio)
+          ensure
+            wio.close
+          end
+        end
+        if compress_threshold
+          chunk = rio.read([DEFLATE_BUFFER_SIZE, compress_threshold].max)
+          return if chunk.bytesize < compress_threshold
+        end
+        deflate(rio, to_io, initial_chunk: chunk)
+      end
+
+      def load(value)
+        rio, wio = binary_pipe
+        within_safe_thread do
+          begin
+            inflate(value, wio)
+          ensure
+            wio.close
+          end
+        end
+        Marshal.load(rio)
+      end
+
+      # Compresses data from IO stream to another one.
+      # Returns number of processed bytes.
+      def deflate(from_io, to_io, initial_chunk: nil)
+        processed_bytes = initial_chunk&.bytesize || 0
+        zlib = Zlib::Deflate.new
+        to_io << zlib.deflate(initial_chunk) if initial_chunk
+        until from_io.eof?
+          chunk = from_io.read(DEFLATE_BUFFER_SIZE)
+          to_io << zlib.deflate(chunk)
+          processed_bytes += chunk.bytesize
+        end
+        to_io << zlib.finish
+        processed_bytes
+      ensure
+        zlib.close
+      end
+
+      def inflate(value, to_io)
+        zlib = Zlib::Inflate.new
+        from_io = StringIO.new(value)
+        to_io << zlib.inflate(from_io.read(INFLATE_CHUNK_SIZE)) until from_io.eof?
+      ensure
+        zlib.close
+      end
+
+      private
+
+        def binary_pipe
+          rio, wio = IO.pipe(binmode: true)
+          wio.set_encoding(Encoding::BINARY)
+          [rio, wio]
```

Comment:
The return here requires an array allocation. Then ruby need another array to do this allocation 

```
rio, wio = binary_pipe
```

(when you assign this way, ruby allocates an array under the hood).

Can we restructure to not allocate those extra two arrays? One alternative could be to have binary_pipe yield those two values:

```ruby
require 'benchmark/ips'

def return_array
  a = "a"
  b = "b"
  return [a, b]
end

def use_block
  a = "a"
  b = "b"
  yield a, b
end

Benchmark.ips do |x|
  x.report("return array") { a, b = return_array }
  x.report("block") { 
    use_block do |a, b|
      # use a, b
    end
  }
  x.compare!
end
# Warming up --------------------------------------
#         return array   216.800k i/100ms
#                block   230.227k i/100ms
# Calculating -------------------------------------
#         return array      4.497M (± 2.1%) i/s -     22.547M in   5.015694s
#                block      5.261M (± 2.3%) i/s -     26.476M in   5.034858s

# Comparison:
#                block:  5261444.1 i/s
#         return array:  4497425.6 i/s - 1.17x  slower
```

Reply:
I'll change. It was using block version of IO.pipe before, but it had seldom strange errors "stream closed" in write. While block was finished after read is finished, which should occur after write finished. Race condition.

Reply:
I think `IO.pipe` will allocate an array for its return value regardless of whether you pass it a block (looking at the source [here](https://ruby-doc.org/core-2.5.0/IO.html#method-c-pipe)).  Therefore, to avoid an extra array allocation, it might be sufficient to do:

```ruby
def binary_pipe
  rw = IO.pipe(binmode: true)
  rw.last.set_encoding(Encoding::BINARY)
  rw
end
```